### ISSUE TITLE
SDL_Log not working on Windows with debugger attached

### ISSUE THREAD

**Issue by namandixit** [libsdl-org/SDL#10523]
After [this commit](https://github.com/libsdl-org/SDL/commit/125ce7137987ec91b931b4fb97d9c3c7af88c377), `SDL_Log` doesn't print anything to the debugger if application is compiled with `/subsystem:windows`. If I use `OutputDebugStringA` or `OutputDebugStringW` manually, it works fine. Reverting to an [older commit](https://github.com/libsdl-org/SDL/commit/ca2c9f680be42c4dfa8c0a0db6ed2f51a567182f) of SDL makes it work again. Setting the `subsystem` to `console` _"works"_, by printing to the `stderr` of attached console instead of the debug stream.

I tested it with Epic's [RAD Debugger](https://github.com/EpicGamesExt/raddebugger/) and [RemedyBG](https://remedybg.itch.io/remedybg) (don't have a Visual Studio install handy, let me know if it is required)

Compilation command was:

    cl.exe main.c /nologo /Zi /FC /source-charset:utf-8 /execution-charset:utf-8 /GS- /std:c11 /DUNICODE=1 /D_UNICODE=1 /Feprogram.exe /link /NOLOGO "libvcruntime.lib" "kernel32.lib" "SDL3.lib" "ucrt.lib" /SUBSYSTEM:WINDOWS /fixed

I was using the callback based `SDL_main`, in case it matters. The commands used to build SDL are [here](https://github.com/mmozeiko/build-sdl3/blob/f06a70002bfe98ae165cb7484263ab331a2ad065/build.cmd).

**Comment by madebr** [comment]
On windows, let's ignore `SDL_LIBC` and always use `WriteFile/WriteConsole` + `OutputDebugStream`.

**Comment by madebr** [comment]
Btw, I can't reproduce with Visual Studio 2019.
I built `testspriteminimal.c` (no callbacks) and `testsprite.c` (callbacks) using the windows and console subsystem, and added these lines:
```c
    SDL_Log("Hello world");
    SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "An error (%s)", argv[0]);
```
```cmd
cl /nologo ..\test\testspriteminimal.c /I../include /Fetestspriteminimal_console.exe /link /NOLOGO "libvcruntime.lib" "kernel32.lib" "SDL3.lib" "ucrt.lib" /SUBSYSTEM:CONSOLE /fixed
cl /nologo ..\test\testspriteminimal.c /I../include /Fetestspriteminimal_windows.exe /link /NOLOGO "libvcruntime.lib" "kernel32.lib" "SDL3.lib" "ucrt.lib" /SUBSYSTEM:WINDOWS /fixed


cl /nologo ..\test\testsprite.c ..\test\testutils.c /I../include /Fetestsprite_console.exe /link /NOLOGO /nodefaultlib:msvcrtd "libvcruntime.lib" "kernel32.lib" "SDL3_test.lib" "SDL3.lib" "ucrt.lib" /SUBSYSTEM:CONSOLE /fixed
cl /nologo ..\test\testsprite.c ..\test\testutils.c /I../include /Fetestsprite_windows.exe /link /NOLOGO /nodefaultlib:msvcrtd "libvcruntime.lib" "kernel32.lib" "SDL3_test.lib" "SDL3.lib" "ucrt.lib" /SUBSYSTEM:WINDOWS /fixed
```

When debugging the resulting executables in the Visual Studio IDE, in both cases I see the following in the debug view (with varying exe path):
```
Hello world
ERROR: An error (C:\projects\SDL\cmake-build-debug\testspriteminimal_console.exe)
```

**Comment by madebr** [comment]
> Regardless, I would say the the older behaviour was far more intuitive and SDL should revert back to it in order to maintain compatibility with non-VS IDEs, debuggers, etc.

Yup, I can reproduce your issue with raddbg.
MSVC forwards stdout/stderr to the debug stream, whereas other debuggers do not.
I don't know how to detect this, so let's log with the win32 api.

We want to keep the logging simple and reliable.

**Comment by slouken** [comment]
@madebr I just ran into this tonight, loading the Visual Studio project in VS2022 yields no debug output and fprintf(stderr) doesn't output anywhere.

We need to fix this... how did it used to work?

**Comment by madebr** [comment]
The change was 125ce7137987ec91b931b4fb97d9c3c7af88c377

<details>

<summary>This patch undoes the failed attempt to detect a MS debugger</summary>

```patch
--- a/src/SDL_log.c
+++ b/src/SDL_log.c
@@ -515,7 +515,6 @@ enum {
     CONSOLE_UNATTACHED = 0,
     CONSOLE_ATTACHED_CONSOLE = 1,
     CONSOLE_ATTACHED_FILE = 2,
-    CONSOLE_ATTACHED_MSVC = 3,
     CONSOLE_ATTACHED_ERROR = -1,
 } consoleAttached = CONSOLE_UNATTACHED;
 
@@ -535,13 +534,11 @@ static void SDLCALL SDL_LogOutput(void *userdata, int category, SDL_LogPriority
         LPTSTR tstr;
         bool isstack;
 
-#if !defined(SDL_PLATFORM_WINRT) && !defined(SDL_PLATFORM_GDK)
+#if !defined(HAVE_STDIO_H) && !defined(SDL_PLATFORM_WINRT) && !defined(SDL_PLATFORM_GDK)
         BOOL attachResult;
         DWORD attachError;
         DWORD consoleMode;
-#if !defined(HAVE_STDIO_H)
         DWORD charsWritten;
-#endif
 
         // Maybe attach console and get stderr handle
         if (consoleAttached == CONSOLE_UNATTACHED) {
@@ -551,7 +548,7 @@ static void SDLCALL SDL_LogOutput(void *userdata, int category, SDL_LogPriority
                 if (attachError == ERROR_INVALID_HANDLE) {
                     // This is expected when running from Visual Studio
                     // OutputDebugString(TEXT("Parent process has no console\r\n"));
-                    consoleAttached = CONSOLE_ATTACHED_MSVC;
+                    consoleAttached = CONSOLE_ATTACHED_ERROR;
                 } else if (attachError == ERROR_GEN_FAILURE) {
                     OutputDebugString(TEXT("Could not attach to console of parent process\r\n"));
                     consoleAttached = CONSOLE_ATTACHED_ERROR;
@@ -582,17 +579,8 @@ static void SDLCALL SDL_LogOutput(void *userdata, int category, SDL_LogPriority
         (void)SDL_snprintf(output, length, "%s%s\r\n", SDL_GetLogPriorityPrefix(priority), message);
         tstr = WIN_UTF8ToString(output);
 
-
-#if defined(HAVE_STDIO_H) && !defined(SDL_PLATFORM_WINRT) && !defined(SDL_PLATFORM_GDK)
-        // When running in MSVC and using stdio, rely on forwarding of stderr to the debug stream
-        if (consoleAttached != CONSOLE_ATTACHED_MSVC) {
-            // Output to debugger
-            OutputDebugString(tstr);
-        }
-#else
         // Output to debugger
         OutputDebugString(tstr);
-#endif
 
 #if !defined(HAVE_STDIO_H) && !defined(SDL_PLATFORM_WINRT) && !defined(SDL_PLATFORM_GDK)
         // Screen output to stderr, if console was attached.
```

</details>

**Comment by madebr** [comment]
This should be fixed now by #10616

Thanks for the report!
